coala_quickstart/generation/Bears.py

```diff
@@ -21,27 +28,79 @@ def filter_relevant_bears(used_languages, arg_parser=None):
     log_printer = LogPrinter(NullPrinter())
     used_languages.append(("All", 100))
 
-    all_bears_by_lang = {
+    bears_by_lang = {
         lang: set(inverse_dicts(*get_filtered_bears([lang],
                                                     log_printer,
                                                     arg_parser)).keys())
         for lang, _ in used_languages
     }
 
-    bears_by_lang = {}
-    for lang in all_bears_by_lang:
-        if lang in IMPORTANT_BEAR_LIST:
-            bears_by_lang[lang] = {bear for bear in all_bears_by_lang[lang]
-                                   if bear.name in IMPORTANT_BEAR_LIST[lang]}
-        else:
-            bears_by_lang[lang] = all_bears_by_lang[lang]
-
     # Each language would also have the language independent bears. We remove
     # those and put them in the "All" category.
-    lang_bears = {lang: bears_by_lang[lang] - bears_by_lang["All"]
-                  for lang, _ in used_languages}
-    lang_bears["All"] = bears_by_lang["All"]
-    return lang_bears
+    all_lang_bears = bears_by_lang["All"]
+    bears_by_lang = {lang: bears_by_lang[lang] - bears_by_lang["All"]
+                     for lang, _ in used_languages}
+    bears_by_lang["All"] = all_lang_bears
+
+    selected_bears = {}
+    candidate_bears = copy.copy(bears_by_lang)
+    to_propose_bears = {}
+
+    # Initialize selected_bears with IMPORTANT_BEAR_LIST
+    for lang, lang_bears in candidate_bears.items():
+        if lang_bears and lang in IMPORTANT_BEAR_LIST:
+            selected_bears[lang] = set()
+            for bear in lang_bears:
+                if bear.__name__ in IMPORTANT_BEAR_LIST[lang]:
+                    selected_bears[lang].add(bear)
+        if lang_bears and lang not in IMPORTANT_BEAR_LIST:
+            selected_bears[lang] = lang_bears
+
+        candidate_bears[lang] = set(
+            [bear for bear in lang_bears
+             if lang in selected_bears and
+             bear not in selected_bears[lang]])
+
+    project_dependency_info = extracted_info.get("ProjectDependencyInfo")
+
+    # Use project_dependency_info to propose bears to user.
+    if project_dependency_info:
+        for lang, lang_bears in candidate_bears.items():
+            matching_dep_bears = get_bears_with_matching_dependencies(
+                lang_bears, project_dependency_info)
+            to_propose_bears[lang] = set(matching_dep_bears)
+            bears_to_remove = [match for match in matching_dep_bears]
```

Comment:
yeah, it's not required here for this commit, probably got mixed up while rebase